backend/src/controllers/UserController.ts

```diff
@@ -155,4 +171,143 @@ async function getPermissions(user: Doc<UserModel>): Promise<Permission[]> {
         .map((permission) => Permission[permission as keyof typeof Permission]);
 }
 
-export { devLoginUser, loginUser, getAllUsers, updateUser, getPermissions };
+/**
+ * @desc 	Gives user roles specified in request body
+ * @route 	POST /api/users/:userId/roles/
+ */
+
+const giveRoles = asyncHandler(async (req: TypedRequest<{ roles: string[] }, UserIdParam>, res: Response) => {
+    // Check atleast one role specified
+    if (req.body.roles.length == 0) {
+        res.status(400).json({
+            status: 'error',
+            message: 'You must specify atleast one role',
+        });
+        return;
+    }
+
+    // Get the user
+    const user = await User.findById(req.params.userId);
+    if (!user) {
+        res.status(400).json({
+            status: 'error',
+            message: 'User not found',
+        });
+        return;
+    }
+
+    const addedRoles: RoleModel[] = [];
+
+    for (const roleId of req.body.roles) {
+        // Check if role exists
+        const role = await Role.findById(roleId);
+        if (!role) {
+            res.status(400).json({
+                status: 'error',
+                message: 'Role not found',
+            });
+            return;
+        }
+
+        // Check if the user already has the role
+        if (user.roles.includes(role.id)) {
+            res.status(400).json({
+                status: 'error',
+                message: 'User already has that role',
+            });
+            return;
+        }
+
+        // Check if trying to add the same role twice
+        if (addedRoles.includes(role.id)) {
+            res.status(400).json({
+                status: 'error',
+                message: 'You cannot add the same role twice',
+            });
+            return;
+        }
+
+        addedRoles.push(role);
+    }
+
+    // Give the roles to the user
+    for (const role of addedRoles) {
+        user.roles.push(role);
+    }
+
```

Comment:
I'm not sure if the fact that they're mongoose objects has any impact on this, but can you do:
```ts
user.roles = user.roles.concat(addedRoles);
```